src/main/java/uk/ac/ebi/atlas/search/metadata/MetadataJsonSerializer.java

```diff
@@ -0,0 +1,173 @@
+
+        //experiment accessions by species
+        var experimentAccessionsBySpecies = cellIdsByExperimentAccession.keySet()
+                .stream()
+                .collect(groupingBy(experimentAccession ->
+                        experimentTrader.getExperiment(experimentAccession, accessKey).getSpecies().getName()));
+
+        var cellTypes = new HashMap<>();
+
+        for (var species : experimentAccessionsBySpecies.keySet()) {
+            var cellTypesBySpecies = new HashMap<>();
+            for (var experimentAccession : experimentAccessionsBySpecies.get(species)) {
+                var cellTypesByExperiment = new ArrayList<String>();
+                for (var cellId : cellIdsByExperimentAccession.get(experimentAccession).keySet()) {
+                    var cellType = cellIdsByExperimentAccession.get(experimentAccession).get(cellId);
+                    cellTypesByExperiment.add(cellType);
+                }
+                cellTypesBySpecies.put(experimentAccession, cellTypesByExperiment.stream().distinct().collect(Collectors.toList()));
+            }
+            cellTypes.put(species, cellTypesBySpecies);
+        }
+
+        return GSON.toJson(
+                ImmutableMap.of("cellTypes", cellTypes)
+        );
+    }
+
+    @Cacheable(cacheNames = "jsonCellTypeMetadata", key = "{#characteristicName, #characteristicValue}")
+    public String cellTypeExpressions(String characteristicName, String characteristicValue, String accessKey) {
+
+        //cell ids
+        var cellIdsByExperimentAccession = fetchCellTypeMetadata(characteristicName, characteristicValue);
+
+        //experiment accessions by species
+        var experimentAccessionsBySpecies = cellIdsByExperimentAccession.keySet()
+                .stream()
+                .collect(groupingBy(experimentAccession ->
+                        experimentTrader.getExperiment(experimentAccession, accessKey).getSpecies().getName()));
+
+        //marker genes by species
+        var markerGenesBySpecies = experimentAccessionsBySpecies
+                .entrySet()
+                .stream()
+                .collect(toMap(
+                        Map.Entry::getKey,
+                        entry ->  entry.getValue().stream()
+                                .flatMap(
+                                        accession -> markerGenesDao.getMarkerGenesWithAveragesPerCluster(accession,
+                                                tsnePlotSettingsService.getExpectedClusters(accession).orElse(10))
+                                                .stream()
+                                                .map(makergene -> makergene.geneId())
+
+                                ).collect(Collectors.toList())
+                        )
+                );
+
+        var expressionByExpressionWithCellType = new HashMap<>();
+
+        for (var species : experimentAccessionsBySpecies.keySet()) {
```

Comment:
Can we use `streams` instead of `for` loops here? The time complexity is around O(n^4) for this code which would affect the performance of the code greatly.